public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts

```diff
@@ -203,7 +203,10 @@ export default class Datasource extends DataSourceApi<AzureMonitorQuery, AzureDa
       }
 
       const ds = this.pseudoDatasource[query.queryType];
-      return ds?.applyTemplateVariables(query, scopedVars) ?? query;
+      return {
+        datasource: ds?.getRef(),
```

Comment:
I think that a better place for doing this is in the `query` method (that we know for sure it will be executed in all the cases). In the loop that starts at line 71 we are already doing some modifications to the query targets, we can se the datasource ref there.

Also, this file has no coverage, it's a good moment to start adding some tests and create the necessary mocks and test structures :)

Reply:
`query` doesn't get executed for recorded queries.

`DataSourceWithBackend`'s `query` already sets the datasource ref https://github.com/grafana/grafana/blob/1d7d8bbf960a77c9784dccf664f844f8d9d7bf80/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L144

Reply:
This relates to my PR (#49642), in that PR I've spread all original properties of the query. I wonder if that change should be made here instead, thoughts?

Reply:
Looking at the description from https://github.com/grafana/grafana/issues/49641

> Azure monitor queries that are referenced in an expression query fail as the datasource is not set (the same would also be true for a log query or resource graph query).

It sounds like this PR would fix that as well. Is that correct, or do you need the other properties of the query as well?

Reply:
Right, I can confirm that this also fixes #49641

Reply:
It would, my question is more are there any other query properties that should be here? Or should we be spreading the whole original query and only overriding the properties as needed. So it'd be something like:
```
return {
   ...query,
   ...(ds?.applyTemplateVariables(query, scopedVars) ?? query),
}

Reply:
> `query` doesn't get executed for recorded queries.

That's surprising :thinking: 

> It would, my question is more are there any other query properties that should be here? Or should we be spreading the whole original query and only overriding the properties as needed. So it'd be something like:
> 
> ```
> return {
>    ...query,
>    ...(ds?.applyTemplateVariables(query, scopedVars) ?? query),
> }
> ```

Yeah, the problem is that `applyTemplateVariables` is removing some properties from `query` which is shouldn't IMO. You fixed that in your PR so I would merge both, even though just this one is enough to fix the issue.